src/tree.rs

```diff
@@ -660,6 +661,30 @@ impl<T> Node<T> {
     }
 }
 
+impl<T> Node<T> {
+    /// Iterates over the tree and calls the given visitor function
+    /// with fully resolved path and its value.
+    pub fn for_each<V: FnMut(String, T) -> bool>(&self, mut visitor: V) {
+        let mut queue = VecDeque::new();
+        queue.push_back((self.prefix.clone(), self));
+        while let Some((mut prefix, node)) = queue.pop_front() {
+            denormalize_params(&mut prefix, &node.remapping);
+            if node.value.is_some() {
+                let path = String::from_utf8(prefix.unescaped().to_vec()).unwrap();
+                let value = unsafe { node.value.as_ref().map(|x| std::ptr::read(x.get())) };
```

Comment:
Hmm we shouldnt be using `ptr::read` here, this could cause double frees. You should instead change the visitor to take `&T` and pass `&*value.get()` here.

Reply:
@ibraheemdev good catch, thanks. But `insert` requires `T` and not `&T`.

Reply:
@ibraheemdev just committed 2nd option to keep zero-copy,

Reply:
Ah right. Consuming the child router makes more sense then.